### ISSUE TITLE
TreeExplainer for XGBRanker and LGBMRanker

### ISSUE THREAD

**Issue by tdyku** [shap/shap#531]
`shap.TreeExplainer` has great support for all kinds of trees. However there are 2 classes in `xgboost` and `lightgbm` which are not supported for now:
* `xgboost.XGBRanker`
* `lightgbm.LGBMRanker`

Is it possible to make it work? Maybe it could be just handled in the same way as regression?

**Comment by annaveronika** [comment]
@tdyczek I'm from catboost team. I'm sorry for interrupting inside a shap issue. And it surely is not the best place to promote some library, but I strongly recommend to try catboost, if you do ranking. I will be very much surprised if you will not get quality boost out of it.

**Comment by slundberg** [comment]
@tdyczek `xgboost.XGBRanker` and `lightgbm.LGBMRanker` are just wrappers around the core api using a ranking objective. So you could also use the original (non-sklearn) APIs. That being said, these should also be supported so I am adding a help wanted tag :) ...they just need to be handled in the TreeEnsemble constructor in a way similar to the regression methods. An example for LightGBM: https://github.com/slundberg/shap/blob/dc936d0497ae37ae223e8b4287c786eb10959cd4/shap/explainers/tree.py#L522-L535

As for the various performances of the libraries on ranking I won't comment, but @annaveronika I think CatBoost already supports SHAP values for ranking right?

**Comment by imatiach-msft** [comment]
@slundberg I was looking at LGBMRanker, the objective is lambdarank.  Probably getting 'margin' output should be straightforward but probability/logloss wouldn't make sense, is that correct?  Would that mean adding an NDCG model_output?  That would probably also involve changing the native layer as it looks like only these output transforms are allowed:

output_transform_codes = {
    "identity": 0,
    "logistic": 1,
    "logistic_nlogloss": 2,
    "squared_loss": 3
}
I sent an initial WIP PR here:
https://github.com/slundberg/shap/pull/532
I'm also trying to find a good dataset to test.  I see lightgbm has a dataset here but it's a bit large:
https://raw.githubusercontent.com/Microsoft/LightGBM/master/examples/lambdarank/rank.train
Also, I don't think it conforms to XGBoost's format.

**Comment by imatiach-msft** [comment]
For XGBRanker the example I've found is this guy:
https://github.com/dmlc/xgboost/blob/24a268a2e3cb17302db3d72da8f04016b7d352d9/demo/rank/rank_sklearn.py
which uses the dataset here:
https://s3-us-west-2.amazonaws.com/xgboost-examples/MQ2008.rar
But that also seems a bit large (and I'm not sure about the copyright for it).
@tdyczek do you know of any good toy datasets for ranking?

**Comment by imatiach-msft** [comment]
@annaveronika is there a "catboost.core.CatBoostRanker"?  I couldn't find it when searching.  I found this tutorial for ranking with catboost: 
https://github.com/catboost/tutorials/blob/master/ranking/ranking_tutorial.ipynb
It looks like it doesn't use a scikit-learn style API.  Is there a scikit-learn catboost ranker?
Also, based on the notebook it looks like the learner output is a "catboost.core.CatBoost", if I understand correctly, but how would one know that it is a ranker specifically?

**Comment by annaveronika** [comment]
https://github.com/catboost/catboost/issues/766 - here is an issue in CatBoost, I think we should discuss CatBoost there. Yes SHAP values are supported in CatBoost, and they have an advantage of calculating much faster than in other libraries if depth is more than 4.
As for SHAP values in ranking - you need to be carefull when making conclusions in this modes. In ranking groupwise features might have large shap values, but in the same time have little effect on resulting loss function. This is because resulting formula values might change a lot between groups, which doesn't influence resulting metric.
So you need to not think that groupwise features are the most important ones inspite of the fact that they might have large SHAP.
@slundberg please correct me if I'm wrong.

**Comment by slundberg** [comment]
@imatiach-msft thanks! I think the margin explanations are plenty for now. We can leave the `tree_output` as none and it will be caught by: https://github.com/slundberg/shap/blob/dc936d0497ae37ae223e8b4287c786eb10959cd4/shap/explainers/tree.py#L104

As for datasets, I would suggest just taking a very small subset of one of those. Or using what LightGBM does: https://github.com/Microsoft/LightGBM/blob/master/tests/python_package_test/test_sklearn.py#L55-L63

@annaveronika I am not an expert on groupwise ranking applications, but it seems similar to a Cox loss in that it is relative to other samples, and in that case the explanations still make sense.

**Comment by imatiach-msft** [comment]
please close as this has been fixed with https://github.com/slundberg/shap/pull/532
